README.md
[![Dub version](https://img.shields.io/dub/v/dolina.svg)](https://code.dlang.org/packages/dolina)
[![Build Status](https://travis-ci.org/o3o/dolina.svg?branch=master)](https://travis-ci.org/o3o/dolina)
[![Dub download Status](https://img.shields.io/dub/dt/dolina.svg)](https://code.dlang.org/packages/dolina)


# dolina
D implementation of HostLink, a communication protocol for use with PLC's made by Omron.

## Libraries
* [serial port library](https://github.com/NCrashed/serial-port)
* [unit-threaded](https://github.com/atilaneves/unit-threaded.git)

## Acknowledgement
The [dmrange](https://github.com/o3o/dolina/blob/master/src/dolina/dmrange.d) code is based on [binrange](https://github.com/p0nce/binrange).
Thanks to ponce.

## References
* [harbored-mod A documentation generator](https://github.com/kiith-sa/harbored-mod)
* [Wiki page about HostLink](https://en.wikipedia.org/wiki/HostLink_Protocol)


CHANGELOG.md
# Dolina Changelog

## [Unreleased]

## [0.6.0] - 2017-09-08
- Merge dmrange into util

## [0.5.1] - 2016-08-02
-  Replaced dmock with unit-threaded mock

## [0.5.0] - 2015-02-21
- Removed getFromDM


tests/app.d
//dolina.channel,dolina.hostlink,dolina,dolina.semver,dolina.util,tests.hostlink
//Automatically generated by unit_threaded.gen_ut_main, do not edit by hand.
import unit_threaded;

int main(string[] args)
{
    return args.runTests!(
                          "dolina.channel",
                          "dolina.hostlink",
                          "dolina",
                          "dolina.semver",
                          "dolina.util",
                          "tests.hostlink"
                          );
}


tests/hostlink.d
module tests.hostlink;

import unit_threaded;

import dolina.channel;
import dolina.hostlink;

@UnitTest
void readShouldSendValidString() {
   auto m = mock!IHostLinkChannel;

   m.expect!"write"("@02RD0031000254*\r");
   m.expect!"read";
   m.returnValue!"read"("@02RD0015*");

   auto host = new HostLink(m);
   host.unit = 2;
   host.readDM(31, 2);
   m.verify();
}


@UnitTest
void writeShouldSendValidString() {
   auto chan = mock!IHostLinkChannel;
   chan.expect!"write"("@02WD03020064196458*\r");
   chan.expect!"read";
   chan.returnValue!"read"("@02WD0015*");



   auto host = new HostLink(chan);
   host.unit = 2;
   host.writeDM(302, [100, 6500]);
   chan.verify();
}

@UnitTest
void GivenBigDataWriteShouldSendMultiplePack() {
   auto chan = mock!IHostLinkChannel;

   // non conta cio' che si scrive
   chan.expect!"write";
   chan.expect!"read";
   chan.expect!"write";
   chan.expect!"read";
   chan.expect!"write";
   chan.expect!"read";

   chan.returnValue!"read"(
      "@02WD0015*",
      "@02WD0015*",
      "@02WD0015*"
      );

   auto host = new HostLink(chan);
   host.unit = 2;

   ushort[62] data = 1;
   host.writeDM(302, data);

   chan.verify();
}


src/dolina/util.d
/**
 * Utility functions to convert DM
 */
module dolina.util;
version(unittest) {
   import unit_threaded;
}

import std.bitmanip : read;
import std.system : Endian; // for Endian
import std.range;

/**
 * Bytes per each DM.
 *
 * A DM is a $(Dushort), so each DM has 2 bytes.
 */
enum BYTES_PER_DM = 2;

/**
 * Converts an array of type T into an ubyte array.
 *
 * Omron stores data in LittleEndian format.
 *
 * Params:  input = array of type T to convert
 *
 * Returns: An array of ubyte
 */
ubyte[] toBytes(T)(T[] input) {
   import std.array : appender;
   import std.bitmanip : append;

   auto buffer = appender!(const(ubyte)[])();
   foreach (dm; input) {
      buffer.append!(T, Endian.littleEndian)(dm);
   }
   return buffer.data.dup;
}
///
unittest {
   [0x8034].toBytes!ushort().shouldEqual([0x34, 0x80]);

   ushort[] buf = [0x8034, 0x2010];
   buf.toBytes!ushort().shouldEqual([0x34, 0x80, 0x10, 0x20]);
   buf.length.shouldEqual(2);

   [0x8034].toBytes!uint().shouldEqual([0x34, 0x80, 0, 0]);
   [0x010464].toBytes!uint().shouldEqual([0x64, 0x04, 1, 0]);
}

/**
 * Converts an array of bytes into DM (ushort) array.
 *
 * Params:  bytes = array to convert
 *
 * Returns: DM rapresentation
 */
ushort[] toDM(ubyte[] bytes) {
   ushort[] dm;
   while (bytes.length >= BYTES_PER_DM) {
      dm ~= bytes.read!(ushort, Endian.littleEndian);
   }
   if (bytes.length > 0) {
      dm ~= bytes[0];
   }
   return dm;
}

///
unittest {
   [0x10].toDM().shouldEqual([0x10]);
   [0, 0xAB].toDM().shouldEqual([0xAB00]);
   [0x20, 0x0].toDM().shouldEqual([0x20]);
   [0x10, 0x20, 0x30, 0x40, 0x50].toDM().shouldEqual([0x2010, 0x4030, 0x50]);
}

/**
 * Takes an array of DM $(D ushort) and converts the first $(D T.sizeof / 2)
 * DM to $(D T).
 * The array is **not** consumed.
 *
 * Params:
 *  T = The integral type to convert the first `T.sizeof / 2` words to.
 *  words = The array of DM to convert
 */
T peekDM(T)(ushort[] words) {
   return peekDM!T(words, 0);
}
///
unittest {
   ushort[] words = [0x645A, 0x3ffb];
   words.peekDM!float.shouldEqual(1.964F);
   words.length.shouldEqual(2);

   ushort[] odd = [0x645A, 0x3ffb, 0xffaa];
   odd.peekDM!float.shouldEqual(1.964F);
   odd.length.shouldEqual(3);
}

/**
 * Takes an array of DM ($(D ushort)) and converts the first $(D T.sizeof / 2)
 * DM to $(D T) starting from index *index*.
 *
 * The array is **not** consumed.
 *
 * Params:
 * T = The integral type to convert the first `T.sizeof / 2` DM to.
 * words = The array of DM to convert
 * index = The index to start reading from (instead of starting at the front).
 */
T peekDM(T)(ushort[] words, size_t index) {
   import std.bitmanip : peek;
   ubyte[] buffer = toBytes(words);
   return buffer.peek!(T, Endian.littleEndian)(index * BYTES_PER_DM);
}
///
unittest {
   [0x645A, 0x3ffb].peekDM!float(0).shouldEqual(1.964F);
   [0, 0, 0x645A, 0x3ffb].peekDM!float(2).shouldEqual(1.964F);
   [0, 0, 0x645A, 0x3ffb].peekDM!float(0).shouldEqual(0);
   [0x80, 0, 0].peekDM!ushort(0).shouldEqual(128);
   [0xFFFF].peekDM!short(0).shouldEqual(-1);
   [0xFFFF].peekDM!ushort(0).shouldEqual(65_535);
   [0xFFF7].peekDM!ushort(0).shouldEqual(65_527);
   [0xFFF7].peekDM!short(0).shouldEqual(-9);
   [0xFFFB].peekDM!short(0).shouldEqual(-5);
   [0xFFFB].peekDM!ushort(0).shouldEqual(65_531);
   [0x8000].peekDM!short(0).shouldEqual(-32_768);
}

/**
 * Converts ushort value into BDC format
 *
 * Params:
 *  dec = ushort in decimal format
 *
 * Returns: BCD value
 */
ushort toBCD(ushort dec) {
   enum ushort MAX_VALUE = 9999;
   enum ushort MIN_VALUE = 0;
   if ((dec > MAX_VALUE) || (dec < MIN_VALUE)) {
      throw new Exception("Decimal out of range (should be 0..9999)");
   } else {
      ushort bcd;
      enum ushort NUM_BASE = 10;
      ushort i;
      for(; dec > 0; dec /= NUM_BASE) {
         ushort rem = cast(ushort)(dec % NUM_BASE);
         bcd += cast(ushort)(rem << 4 * i++);
      }
      return bcd;
   }
}
///
unittest {
   0.toBCD().shouldEqual(0);
   10.toBCD().shouldEqual(0x10);
   34.toBCD().shouldEqual(52);
   127.toBCD().shouldEqual(0x127);
   110.toBCD().shouldEqual(0x110);
   9999.toBCD().shouldEqual(0x9999);
   9999.toBCD().shouldEqual(39_321);
}

/**
 * Converts BCD value into decimal format
 *
 * Params:
 *  dec = ushort in BCD format
 *
 * Returns: decimal value
 */
ushort fromBCD(ushort bcd) {
   enum int NO_OF_DIGITS = 8;
   enum ushort MAX_VALUE = 0x9999;
   enum ushort MIN_VALUE = 0;
   if ((bcd > MAX_VALUE) || (bcd < MIN_VALUE)) {
      throw new Exception("BCD out of range (should be 0..39321)");
   } else {
      ushort dec;
      ushort weight = 1;
      foreach (j; 0 .. NO_OF_DIGITS) {
         dec += cast(ushort)((bcd & 0x0F) * weight);
         bcd = cast(ushort)(bcd >> 4);
         weight *= 10;
      }
      return dec;
   }
}
///
unittest {
   0.fromBCD().shouldEqual(0);

   (0x22).fromBCD().shouldEqual(22);
   (34).fromBCD().shouldEqual(22);
   // 17bcd
   (0b0001_0111).fromBCD().shouldEqual(17);
   295.fromBCD().shouldEqual(127);
   39_321.fromBCD().shouldEqual(9_999);
   (0x9999).fromBCD().shouldEqual(9_999);
}

/**
 * Takes an input range of DM ($(D ushort)) and converts the first $(D T.sizeof / 2)
 * DM's to $(D T).
 * The array is consumed.
 *
 * Params:
 * T = The integral type to convert the first `T.sizeof / 2` DM to.
 * input = The input range of DM to convert
 */
T pop(T, R)(ref R input) if ((isInputRange!R)
      && is(ElementType!R : const ushort)) {
   import std.traits : isIntegral, isSigned;
   static if(isIntegral!T) {
      return popInteger!(R, T.sizeof / 2, isSigned!T)(input);
   } else static if (is(T == float)) {
      return uint2float(popInteger!(R, 2, false)(input));
   } else static if (is(T == double)) {
      return ulong2double(popInteger!(R, 4, false)(input));
   } else {
      static assert(false, "Unsupported type " ~ T.stringof);
   }
}

/**
 * $(D pop!float) example
 */
unittest {
   ushort[] asPeek = [0x645A, 0x3ffb];
   asPeek.pop!float.shouldEqual(1.964F);
   asPeek.length.shouldEqual(0);

   // float.sizeOf is 4bytes => 2 DM
   ushort[] input = [0x1eb8, 0xc19d];
   input.length.shouldEqual(2);
   pop!float(input).shouldEqual(-19.64F);
   input.length.shouldEqual(0);

   input = [0x0, 0xBF00, 0x0, 0x3F00];
   input.pop!float.shouldEqual(-0.5F);
   pop!float(input).shouldEqual(0.5F);
}

/**
 * $(D pop!double) examples.
 *
 * A double has size 8 bytes => 4DM
 */
unittest {
   ushort[] input = [0x0, 0x0, 0x0, 0x3FE0];
   input.length.shouldEqual(4);

   pop!double(input).shouldEqual(0.5);
   input.length.shouldEqual(0);

   input = [0x0, 0x0, 0x0, 0xBFE0];
   pop!double(input).shouldEqual(-0.5);

   input = [0x00, 0x01, 0x02, 0x03];
   input.length.shouldEqual(4);
   pop!int(input).shouldEqual(0x10000);
   input.length.shouldEqual(2);
   pop!int(input).shouldEqual(0x30002);
   input.length.shouldEqual(0);
}

/**
 * pop!ushort and short examples
 */
unittest {
   ushort[] input = [0xFFFF, 0xFFFF, 0xFFFB, 0xFFFB];
   pop!ushort(input).shouldEqual(0xFFFF);
   pop!short(input).shouldEqual(-1);
   pop!ushort(input).shouldEqual(0xFFFB);
   pop!short(input).shouldEqual(-5);
} unittest {
   ushort[] input = [0x1eb8, 0xc19d
      , 0x0, 0xBF00, 0x0, 0x3F00
      , 0x0, 0x0, 0x0, 0x3FE0
      , 0x0, 0x0, 0x0, 0xBFE0
      , 0x00, 0x01, 0x02, 0x03
      , 0xFFFF, 0xFFFF, 0xFFFB, 0xFFFB];

   pop!float(input).shouldEqual(-19.64F);
   pop!float(input).shouldEqual(-0.5F);
   pop!float(input).shouldEqual(0.5F);

   pop!double(input).shouldEqual(0.5);
   pop!double(input).shouldEqual(-0.5);

   pop!int(input).shouldEqual(0x10000);
   pop!int(input).shouldEqual(0x30002);

   pop!ushort(input).shouldEqual(0xFFFF);
   pop!short(input).shouldEqual(-1);
   pop!ushort(input).shouldEqual(0xFFFB);
   pop!short(input).shouldEqual(-5);
} unittest {
   ushort[] shortBuffer = [0x645A];
   shortBuffer.length.shouldEqual(1);
   shortBuffer.pop!float().shouldThrow!Exception;

   ushort[] bBuffer = [0x0001, 0x0002, 0x0003];
   bBuffer.length.shouldEqual(3);

   bBuffer.pop!ushort.shouldEqual(1);
   bBuffer.length.shouldEqual(2);

   bBuffer.pop!ushort.shouldEqual(2);
   bBuffer.length.shouldEqual(1);

   bBuffer.pop!ushort.shouldEqual(3);
   bBuffer.length.shouldEqual(0);
}

/**
 * Takes an input range of DM ($(D ushort)) and converts the first $(D numDM)
 * DM's to $(D T).
 * The array is consumed.
 *
 * Params:
 *  R = The integral type of innput range
 *  numDM = Number of DM to convert
 *  wantSigned = Get signed value
 *  input = The input range of DM to convert
 */
private auto popInteger(R, int numDM, bool wantSigned)(ref R input) if ((isInputRange!R)
      && is(ElementType!R : const ushort)) {
   import std.traits : Signed;
   alias T = IntegerLargerThan!(numDM);
   T result = 0;

   foreach (i; 0 .. numDM) {
      result |= ( cast(T)(popDM(input)) << (16 * i) );
   }

   static if (wantSigned) {
      return cast(Signed!T)result;
   }  else {
      return result;
   }
} unittest {
   ushort[] input = [0x00, 0x01, 0x02, 0x03];
   popInteger!(ushort[], 2, false)(input).shouldEqual(0x10000);
   popInteger!(ushort[], 2, false)(input).shouldEqual(0x30002);
   input.length.shouldEqual(0);
   input = [0x01, 0x02, 0x03, 0x04];
   popInteger!(ushort[], 3, false)(input).shouldEqual(0x300020001);

   input = [0x01, 0x02];
   popInteger!(ushort[], 3, false)(input)
      .shouldThrow!Exception;

   input = [0x00, 0x8000];
   popInteger!(ushort[], 2, false)(input).shouldEqual(0x8000_0000);

   input = [0xFFFF, 0xFFFF];
   popInteger!(ushort[], 2, true)(input).shouldEqual(-1);
   input = [0xFFFF, 0xFFFF, 0xFFFB, 0xFFFB];
   popInteger!(ushort[], 1, false)(input).shouldEqual(0xFFFF);
   popInteger!(ushort[], 1, true)(input).shouldEqual(-1);
   popInteger!(ushort[], 1, false)(input).shouldEqual(0xFFFB);
   popInteger!(ushort[], 1, true)(input).shouldEqual(-5);
}

private template IntegerLargerThan(int numDM) if (numDM > 0 && numDM <= 4) {
   static if (numDM == 1) {
      alias IntegerLargerThan = ushort;
   } else static if (numDM == 2) {
      alias IntegerLargerThan = uint;
   } else {
      alias IntegerLargerThan = ulong;
   }
}

private ushort popDM(R)(ref R input) if ((isInputRange!R) && is(ElementType!R : const ushort)) {
   if (input.empty) {
      throw new Exception("Expected a ushort, but found end of input");
   }

   const(ushort) d = input.front;
   input.popFront();
   return d;
}

/**
 * Writes numeric type *T* into a output range of *ushort*.
 *
 * Params:
 * n = The numeric type to write into output range
 * output = The output range of DM to convert
 *
 * Examples:
 * --------------------
 * ushort[] arr;
 * auto app = appender(arr);
 * write!float(app, 1.0f);
 *
 *
 * auto app = appender!(const(ushort)[]);
 * app.write!ushort(5);
 * app.data.shouldEqual([5]);
 * --------------------
 */
void write(T, R)(ref R output, T n) if (isOutputRange!(R, ushort)) {
   import std.traits : isIntegral;
   static if (isIntegral!T) {
      writeInteger!(R, T.sizeof / 2)(output, n);
   } else static if (is(T == float)) {
      writeInteger!(R, 2)(output, float2uint(n));
   } else static if (is(T == double)) {
      writeInteger!(R, 4)(output, double2ulong(n));
   } else {
      static assert(false, "Unsupported type " ~ T.stringof);
   }
}
/**
 * Write float and double
 */
unittest {
   ushort[] arr;
   auto app = appender(arr);
   write!float(app, 1.0f);

   app.write!double(2.0);

   ushort[] expected = [
      0, 0x3f80,
      0, 0, 0, 0x4000];
   app.data.shouldEqual(expected);
}

/**
 * Write ushort and int
 */
unittest {
   import std.array : appender;

   auto app = appender!(const(ushort)[]);
   app.write!ushort(5);
   app.data.shouldEqual([5]);

   app.write!float(1.964F);
   app.data.shouldEqual([5, 0x645A, 0x3ffb]);

   app.write!uint(0x1720_8034);
   app.data.shouldEqual([5, 0x645A, 0x3ffb, 0x8034, 0x1720]);
}

private void writeInteger(R, int numDM)(ref R output, IntegerLargerThan!numDM n) if (isOutputRange!(R, ushort)) {
   import std.traits : Unsigned;
   alias T = IntegerLargerThan!numDM;
   auto u = cast(Unsigned!T)n;
   foreach (i; 0 .. numDM) {
      immutable(ushort) b = (u >> (i * 16)) & 0xFFFF;
      output.put(b);
   }
}

private float uint2float(uint x) pure nothrow {
   float_uint fi;
   fi.i = x;
   return fi.f;
} unittest {
   // see http://gregstoll.dyndns.org/~gregstoll/floattohex/
   uint2float(0x24369620).shouldEqual(3.959212E-17F);
   uint2float(0x3F000000).shouldEqual(0.5F);
   uint2float(0xBF000000).shouldEqual(-0.5F);
   uint2float(0x0).shouldEqual(0);
   uint2float(0x419D1EB8).shouldEqual(19.64F);
   uint2float(0xC19D1EB8).shouldEqual(-19.64F);
   uint2float(0x358637bd).shouldEqual(0.000001F);
   uint2float(0xb58637bd).shouldEqual(-0.000001F);
}

private uint float2uint(float x) pure nothrow {
   float_uint fi;
   fi.f = x;
   return fi.i;
} unittest {
   // see http://gregstoll.dyndns.org/~gregstoll/floattohex/
   float2uint(3.959212E-17F).shouldEqual(0x24369620);
   float2uint(.5F).shouldEqual(0x3F000000);
   float2uint(-.5F).shouldEqual(0xBF000000);
   float2uint(0x0).shouldEqual(0);
   float2uint(19.64F).shouldEqual(0x419D1EB8);
   float2uint(-19.64F).shouldEqual(0xC19D1EB8);
   float2uint(0.000001F).shouldEqual(0x358637bd);
   float2uint(-0.000001F).shouldEqual(0xb58637bd);
}

// read/write 64-bits float
private union float_uint {
   float f;
   uint i;
}

double ulong2double(ulong x) pure nothrow {
   double_ulong fi;
   fi.i = x;
   return fi.f;
} unittest {
   // see http://gregstoll.dyndns.org/~gregstoll/floattohex/
   ulong2double(0x0).shouldEqual(0);
   ulong2double(0x3fe0000000000000).shouldEqual(0.5);
   ulong2double(0xbfe0000000000000).shouldEqual(-0.5);
}

private ulong double2ulong(double x) pure nothrow {
   double_ulong fi;
   fi.f = x;
   return fi.i;
} unittest {
   // see http://gregstoll.dyndns.org/~gregstoll/floattohex/
   double2ulong(0).shouldEqual(0);
   double2ulong(0.5).shouldEqual(0x3fe0000000000000);
   double2ulong(-0.5).shouldEqual(0xbfe0000000000000);
}

private union double_ulong {
   double f;
   ulong i;
}


src/dolina/channel.d
/**
 * Defines channel objects layer.
 *
 * A channel is a message delivery mechanism that forwards a message from a
 * sender to one receiver.
 */
module dolina.channel;

import std.exception;
import serial.device;
/**
 * Defines a basic HostLink channel
 */
interface IHostLinkChannel {
   /**
    * Reads a message from channel
    *
    * Returns: message read
    */
   string read();

   /**
    * Writes a message on channel
    *
    * Params:  message = The message being sent on the channel.
    */
   void write(string message);
}

/**
 * Channel based on serial RS232 communication
 */
class HostLinkChannel: IHostLinkChannel {
   private SerialPort serialPort;
   this(SerialPort serialPort) {
      enforce(serialPort !is null);
      this.serialPort = serialPort;
   }

   string read() {
      enum START = 0x40; // @
      enum END = 0x0D;  // CR
      bool inside;

      ubyte[1] buffer;
      ubyte[] reply;
      ubyte b;
      do {
         immutable(size_t) length = serialPort.read(buffer);
         if (length > 0) {
            b = buffer[0];
            if (b == START) {
               inside = true;
            }
            if (inside) {
               reply ~= b;
            }
         }

      } while (b != END);
      return cast(string)(reply).idup;
   }

   void write(string message) {
      serialPort.write(cast(void[])message);
   }
}


src/dolina/hostlink.d
/**
 * Omron HostLink protocols
 */
module dolina.hostlink;

version(unittest) {
   import unit_threaded;
}

import std.array;
import std.conv;
import std.exception;
import std.stdio;
import std.string;

import dolina.channel;

/**
 * Defines the basic HostLink protocol interface.
 */
interface IHostLink {
   /**
    * Reads the contents of specified DM $(D ushort), starting from the specified
    * address.
    *
    * Params:
    * address = Beginning DM address
    * length = Number of DM to read
    */
   ushort[] readDM(const(int) address, const(int) length);

   /**
    * Writes data to the DM area, starting from the specified address.
    *
    * Params:
    * address = Beginning DM address
    * data = DM ($(D ushort)) to write
    *
    * Examples:
    * --------------------
    * auto buffer = appender!(const(ushort)[]);
    * buffer.write!float(12.68);
    * buffer.write!ushort(5);
    * plc.writeDM(address, buffer.data);
    * --------------------
    */
   void writeDM(const(int) address, const(ushort)[] data);

   /**
    * PLC Unit number.
    */
   @property int unit();
   @property void unit(int u);

   /**
    * Data memory area size (DM)
    */
   @property int dataMemorySize();
}

/**
 * Provides the implementation for HostLink protocol
 */
class HostLink: IHostLink {
   private IHostLinkChannel channel;

   this(IHostLinkChannel channel) {
      assert(channel !is null);
      this.channel = channel;
   }

   @property int dataMemorySize() {
      enum DM_SIZE = 6656;
      return DM_SIZE;
   }

   private int _unit;
   @property int unit() {
      return _unit;
   }

   @property void unit(int u) {
      assert(u >= 0 && u < 32);
      this._unit = u;
   }

   private enum MAX_FRAME_SIZE = 29;
   ushort[] readDM(const(int) address, const(int) length)
      in {
         assert(address >= 0, "negative address");
         assert(length >= 0, "negative lenght");
      } body {
         if (length > MAX_FRAME_SIZE) {
            return readDM(address, MAX_FRAME_SIZE) ~ readDM(address + MAX_FRAME_SIZE, length - MAX_FRAME_SIZE);
         } else if (length <= 0) {
            return null;
         } else {
            string w = getRDString(_unit, address, length);
            channel.write(w);

            string r = channel.read();
            string reply = crop(r);

            immutable(int) err = getErrorCodeFromReply(reply);
            if (err > 0) {
               throw new OmronException(err);
            } else {
               return toDM(reply[6 .. $ - 2]);
            }
         }
      }

   void writeDM(const(int) address, const(ushort)[] data)
      in {
         assert(address >= 0, "negative address");
      } body {
         if (data.length > MAX_FRAME_SIZE) {
            writeDM(address, data[0 .. MAX_FRAME_SIZE]);
            writeDM(address + MAX_FRAME_SIZE, data[MAX_FRAME_SIZE .. $]);
         } else {
            channel.write(getWDString(_unit, address, data));
            string reply = crop(channel.read());

            immutable(int) err = getErrorCodeFromReply(reply);
            if (err > 0) {
               throw new OmronException(err);
            }
         }
      }
}


/**
 * The NullHostLing will not read or write.
 */
class NullHostLink: IHostLink {
   ushort[] readDM(const(int) address, const(int) length) {
      return new ushort[length];
   }

   void writeDM(const(int) address, const(ushort)[] data) { }

   private int _unit;
   @property int unit() {
      return _unit;
   }
   @property void unit(int u) {
      _unit = u;
   }

   @property int dataMemorySize() { return 100; }
}

/*
 * Get a string command to read data memory.
 *
 *  String shold be (see chap. 4.9 Omron manual 1E66)
 * ---
 * @|unit|RD|start|length|fcs|*CR
 * ---
 *
 * Params:
 * unit = Node number
 *
 * address = Beginning word address
 * length = Number of words
 *
 * Returns: RD command string
 */
private pure string getRDString(const(int) unit, const(int) address, const(int) length) {
   string send = "@"
      ~ format("%.2d", unit)
      ~ "RD"
      ~ format("%.4d", address)
      ~ format("%.4d", length);

   send ~= fcs(send) ~ "*\r";
   return send;
} unittest {
   getRDString(0, 23, 5).shouldEqual("@00RD0023000552*\r");
   getRDString(0, 100, 2).shouldEqual("@00RD0100000255*\r");
   getRDString(0, 258, 2).shouldEqual("@00RD025800025B*\r");
}

/*
 * Get a string command to write data memory.
 *
 * String shold be (see chap. 4.22 Omron manual 1E66)
 * ---
 * @|unit|WD|address|d0|d1...|fcs|*CR|
 * ---
 *
 * Params:
 * unit = Node number
 * address = Beginning word address
 * length = Number of words
 * data = Words to write
 *
 * Returns: WD command string
 */
private string getWDString(const(int) unit, const(int) address, const(ushort)[] data) {
   string send = "@" ~ format("%.2d", unit) ~ "WD" ~ format("%.4d", address);
   foreach (i; 0 .. data.length) {
      send ~= format("%.4x", data[i]);
   }
   send ~= fcs(send) ~ "*\r";
   return send;
} unittest {
   getWDString(2, 302, [100, 6500]).shouldEqual("@02WD03020064196458*\r");
}

/*
 * Converts a string returned by DM read into an array of `ushort`.
 *
 * The string consists of groups of four characters, each representing a digit in hex.
 * Each group of four characters is the value of a DM
 *
 * Eg. the string representation of `[15, 16, 17]` is  `000F 0010 0011` (spaces are added for readability)
 *
 * Params:  input = string to convert
 * Returns:  Array of converted values
 */
private ushort[] toDM(string input) {
   // std.array.appender:
   // Convenience function that returns an Appender!A object initialized with
   // array.

   auto data = appender!(ushort[])();
   while(input.length > 0) {
      immutable(ushort) x = parseFront(input);
      data.put(x);
      input = input.length >= 4 ? input[4..$] : [];
   }
   return data.data;
} unittest {
   ushort[][string] testCase = [
      "000100020003": [0x1, 0x2, 0x3],
      "00FF8000FFFF": [0xFF, 0x8000, 0xFFFF],
      "5": [0x5],
      "12345": [0x1234, 0x5],
   ];

   foreach (key, value; testCase) {
      toDM(key).shouldEqual(value);
   }
   toDM("").length.shouldEqual(0);
}

/*
 * Converts a DM string representation into a numeric value $(D_CODE ushort).
 *
 * A DM is represented by four characters: if the input string contains more, then
 * `parseFront` converts the first four, if it contains less then all available characters are converted
 *
 * Params:  input = string to convert
 * Returns:  Converted value
 */
private pure ushort parseFront(string input) {
   assert(input.length > 0);

   string front;
   if (input.length >= 4) {
      front = input[0..4];
   } else {
      front = input[0..$];
   }
   // std.conv
   // 16 indica che la stringa e' hex
   return parse!ushort(front, 16);
} unittest {
   parseFront("000F").shouldEqual(15);
   parseFront("000F00FF").shouldEqual(15);
   parseFront("000Fasd").shouldEqual(15);
   parseFront("F").shouldEqual(15);
   parseFront("0F").shouldEqual(15);
   parseFront("0010").shouldEqual(16);
   parseFront("1").shouldEqual(1);
}

/*
 * Computes FCS of message.
 *
 * The FCS is an 8-bit data represented by two ASCII characters (00 to FF).
 *
 * It is a result of Exclusive OR sequentially performed on each character in
 * the message.
 *
 * Params:  msg = Message
 *
 * Returns: Mesage FCS
 */
private pure string fcs(string msg) {
   if (msg.length == 0) {
      return "00";
   } else {
      /*
         std.string.representation:
         Returns the representation of a string, which has the same
         type as the string except the character type is replaced by ubyte,
         ushort, or uint depending on the character width
       */
      immutable(ubyte[]) b = representation(msg);
      int fcs;
      foreach (i; 0 .. b.length) {
         fcs ^= b[i];
      }
      return std.string.format("%.2X", fcs);
   }
} unittest {
   fcs("@02WD00").shouldEqual("51");
   fcs("12").shouldEqual("03");
   fcs("123").shouldEqual("30");
   fcs("0").shouldEqual("30");
   fcs("abc").shouldEqual("60");
   fcs("@00RD00300001").shouldEqual("54");
   fcs("@00RD00300001").shouldEqual("54");
   fcs("@00RD02580002").shouldEqual("5B");
   fcs("").shouldEqual("00");
   fcs(null).shouldEqual("00");
}

/*
 * Get message data.
 *
 * The message data are between the character '@' at the start and '*' at the end.
 *  ---
 *  @|uu|RD|ee|--data --|fc|*|CR
 *  ---
 *  Params:  message = Input message
 *
 *  Returns:
 * The message data if the message is valid, otherwise an empty string
 */
private string crop(string message) {
   import std.regex : match, regex;
   string data;
   if (message.length > 0) {
      string pattern = r"@(?P<core>[^\*]+)\*.*";
      auto m = match(message, regex(pattern));
      if (m) {
         data = m.captures["core"];
      }
   }
   return data;
} unittest {
   crop("@abc*").shouldEqual("abc");
   crop("@abc").shouldEqual("");
   crop("").shouldEqual("");
   crop("xy").shouldEqual("");
   crop("@01RD00").shouldEqual("");
   crop("@01RD00*").shouldEqual("01RD00");
   crop("@01RD00*\n").shouldEqual("01RD00");
   crop("@01RD00*\n@03").shouldEqual("01RD00");
   crop("@00RD000000000056*\r").shouldEqual("00RD000000000056");
}

/*
 * Returns the error code inside the response.
 *
 * In response, there are eight frame characters around error code:
 *
 * ---
 * uu|RD|ee|...dati..|FC|
 * ---
 * with:
 * $(UL
 * $(LI uu: unit no)
 * $(LI RD: read dm)
 * $(LI ee: error code)
 * )
 *
 * Params: reply = PLC reply
 *
 * Returns: error code or 0 if there are no errors.
 */
private int getErrorCodeFromReply(string reply) {
   int err = 1001;
   writeln("errcode imp: ", reply);

   if (reply.length > 5) {
      string code = reply[4..6];
      writeln("errcode: ", code);
      err = parse!int(code, 16);
   }
   return err;
} unittest {
   getErrorCodeFromReply("02RD0015").shouldEqual(0);
   getErrorCodeFromReply("01RD0").shouldEqual(1001);
   getErrorCodeFromReply("01RD0715").shouldEqual(7);
   getErrorCodeFromReply("01RD0A15").shouldEqual(10);
   getErrorCodeFromReply("01RD0B312").shouldEqual(11);
   getErrorCodeFromReply("").shouldEqual(1001);
   getErrorCodeFromReply("01RD").shouldEqual(1001);
   getErrorCodeFromReply("00RD1354").shouldEqual(0x13);
}

class OmronException: Exception {
   this(int errCode) {
      _errCode = errCode;
      super(getErrorMessage(errCode));
   }
   private int _errCode;
   @property int errCode() {
      return _errCode;
   }
}

private pure string getErrorMessage(const(int) errorNr) {
   switch (errorNr) {
      case 0x00: return "Success";
      case 0x01: return "Execution was not possible the PLC is in RUN mode";
      case 0x13: return "Check sum error";
      case 0x14: return "Command format error";
      case 0x15: return "An incorrect data area designation was made for READ or WRITE";
      case 0x18: return "Frame length error";
      case 0x22: return "The specified memory unit does not exists";
      case 0x23: return "The specified memory unit is write protected";
      case 0xA3: return "Aborted due to checksum error in transmit data";
      case 0xA5: return "Aborted due to entry number data error in transmit data";
      case 0xA6: return "Aborted due to frame length error in transmit data";
      case 0x3E8: return "dolina: Frame lenght error in received data"; //10000
      case 0x3E9: return "dolina: Invalid format of error code"; // 1001
      default: return "Unkown error code " ~ to!string(errorNr);
   }
} unittest {
   import std.string : startsWith;
   getErrorMessage(0x44).startsWith("Unkown").shouldBeTrue;
   getErrorMessage(0x13).shouldEqual("Check sum error");
}
